cardano-db-sync/src/Cardano/DbSync/Era/Byron/Genesis.hs

```diff
@@ -171,52 +175,97 @@ validateGenesisDistribution prunes disInOut tracer networkName cfg bid =
-            , DB.txOutDataHash = Nothing
-            , DB.txOutInlineDatumId = Nothing
-            , DB.txOutReferenceScriptId = Nothing
+      txId <- do
+        DB.insertTx $
+          DB.Tx
+            { DB.txHash = Byron.unTxHash val
+            , DB.txBlockId = blkId
+            , DB.txBlockIndex = 0
+            , DB.txOutSum = DB.DbLovelace (Byron.unsafeGetLovelace value)
+            , DB.txFee = DB.DbLovelace 0
+            , DB.txDeposit = Just 0
+            , DB.txSize = 0 -- Genesis distribution address to not have a size.
+            , DB.txInvalidHereafter = Nothing
+            , DB.txInvalidBefore = Nothing
+            , DB.txValidContract = True
+            , DB.txScriptSize = 0
+            , DB.txTreasuryDonation = DB.DbLovelace 0
             }
+      --
+      unless disInOut $
+        case getTxOutTableType syncEnv of
+          DB.TxOutCore ->
+            void . DB.insertTxOut $
+              DB.CTxOutW
+                C.TxOut
+                  { C.txOutTxId = txId
+                  , C.txOutIndex = 0
+                  , C.txOutAddress = Text.decodeUtf8 $ Byron.addrToBase58 address
+                  , C.txOutAddressHasScript = False
+                  , C.txOutPaymentCred = Nothing
+                  , C.txOutStakeAddressId = Nothing
+                  , C.txOutValue = DB.DbLovelace (Byron.unsafeGetLovelace value)
+                  , C.txOutDataHash = Nothing
+                  , C.txOutInlineDatumId = Nothing
+                  , C.txOutReferenceScriptId = Nothing
+                  , C.txOutConsumedByTxId = Nothing
+                  }
+          DB.TxOutVariantAddress -> do
+            let addrRaw = serialize' address
+                vAddress = mkVAddress addrRaw
+            addrDetailId <- insertAddress addrRaw vAddress
+            void . DB.insertTxOut $
+              DB.VTxOutW (mkVTxOut txId addrDetailId) Nothing
+  where
+    mkVTxOut :: DB.TxId -> V.AddressId -> V.TxOut
+    mkVTxOut txId addrDetailId =
+      V.TxOut
+        { V.txOutTxId = txId
+        , V.txOutIndex = 0
+        , V.txOutValue = DB.DbLovelace (Byron.unsafeGetLovelace value)
+        , V.txOutDataHash = Nothing
+        , V.txOutInlineDatumId = Nothing
+        , V.txOutReferenceScriptId = Nothing
+        , V.txOutAddressId = addrDetailId
+        , V.txOutConsumedByTxId = Nothing
+        }
+
+    mkVAddress :: ByteString -> V.Address
+    mkVAddress addrRaw = do
+      V.Address
+        { V.addressAddress = Text.decodeUtf8 $ Byron.addrToBase58 address
+        , V.addressRaw = addrRaw
+        , V.addressHasScript = False
+        , V.addressPaymentCred = Nothing -- Byron does not have a payment credential.
+        , V.addressStakeAddressId = Nothing -- Byron does not have a stake address.
+        }
+
+    insertAddress ::
+      (MonadBaseControl IO m, MonadIO m) =>
+      ByteString ->
+      V.Address ->
+      ReaderT SqlBackend m V.AddressId
+    insertAddress addrRaw vAdrs = do
```

Comment:
I think doing a query and then an insert is not the best approach, as it does 2 db operations normally. We could use instead `insertCheckUnique`, which does a single query. This will require a Unique Key in the address.raw field, both in the persistent schema and the raw sql that actually create the schema.

Given that unique key, the index on address.raw then becomes reduandant

Reply:
Actually I take that back. The current index causes different problems. Syncing mainnet I got this crash
```
[db-sync-node:Error:85] [2024-09-26 15:16:45.43 UTC] runDBThread: DbInsertException "insertAddress" 
(SqlError {sqlState = "54000", sqlExecStatus = FatalError, sqlErrorMsg = "index row size 7704 exceeds
 btree version 4 maximum 2704 for index \"idx_address_raw\"", sqlErrorDetail = "Index row references
 tuple (32866,36) in relation \"address\".", sqlErrorHint = "Values larger than 1/3 of a buffer page cannot
 be indexed.\nConsider a function index of an MD5 hash of the value, or use full text indexing."})
```
So a unique key will suffer from the same issue. Since [postgres doesn't support unique keys based on hash](https://stackoverflow.com/questions/44274080/postgres-hash-index-with-unique-constraint), the best approach is to replace the current btree index with a hash index